core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java

```diff
@@ -885,31 +901,41 @@ public boolean decide(String workflowId) {
      * @throws ApplicationException if the workflow cannot be paused
      */
     public void pauseWorkflow(String workflowId) {
-        WorkflowStatus status = WorkflowStatus.PAUSED;
-        Workflow workflow = executionDAOFacade.getWorkflowById(workflowId, false);
-        if (workflow.getStatus().isTerminal()) {
-            throw new ApplicationException(CONFLICT, "Workflow id " + workflowId + " has ended, status cannot be updated.");
-        }
-        if (workflow.getStatus().equals(status)) {
-            return;        //Already paused!
+        try {
+            waitForLock(workflowId);
+            WorkflowStatus status = WorkflowStatus.PAUSED;
+            Workflow workflow = executionDAOFacade.getWorkflowById(workflowId, false);
+            if (workflow.getStatus().isTerminal()) {
+                throw new ApplicationException(CONFLICT, "Workflow id " + workflowId + " has ended, status cannot be updated.");
+            }
+            if (workflow.getStatus().equals(status)) {
+                return;        //Already paused!
+            }
+            workflow.setStatus(status);
+            executionDAOFacade.updateWorkflow(workflow);
+        } finally {
+            releaseLock(workflowId);
         }
-        workflow.setStatus(status);
-        executionDAOFacade.updateWorkflow(workflow);
     }
 
     /**
      * @param workflowId
      * @throws IllegalStateException
      */
     public void resumeWorkflow(String workflowId) {
-        Workflow workflow = executionDAOFacade.getWorkflowById(workflowId, false);
-        if (!workflow.getStatus().equals(WorkflowStatus.PAUSED)) {
-            throw new IllegalStateException("The workflow " + workflowId + " is not PAUSED so cannot resume. " +
-                    "Current status is " + workflow.getStatus().name());
+        try {
+            waitForLock(workflowId);
```

Comment:
Consider having a function to construct the lockId which takes the workflowId and the context into account. In the future, we might want to lock workflows for other critical sections also.

Reply:
these are meant as helper methods specifically for execution.  If we want to lock for other reasons, a different string would be passed in (to lock API) as lockId.  Will see if it makes sense to make these methods more general.

Reply:
I understand that are specific to workflow, I was suggesting that if we want to lock the workflow for execution and workflow for completion then it would be a problem. We can consider having a function like :

`
String getWorkflowLockId(String workflowId, String lockContext) {
                return getHash(workflowId, lockContext);
}

waitForLock(getWorkflowLockId("asdxxjhkadkshd", LOCKED_FOR_EXECUTION));
waitForLock(getWorkflowLockId("asdxxjhkadkshd", LOCKED_FOR_COMPLETION));
`

Reply:
@mudit3774 workflowCompletion is part of execution, and would use the same lock.  Happy to continue this conversation, I'm merging this PR first to move forward.